src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,248 @@
+            for (z = current_max - 1; z >= cur; z--)
+                results[z + 1] = results[z];
+            if (++current_max > threshold)
+                current_max = threshold;
+
+            //insert element into the given position
+            results[cur] = i;
+        }
+    }
+}
+
+template <bool is_plain_column, typename T, bool is_weighted>
+class AggregateFunctionGroupSortedArray : public IAggregateFunctionDataHelper<
+                                              AggregateFunctionGroupSortedArrayData<T, is_weighted>,
+                                              AggregateFunctionGroupSortedArray<is_plain_column, T, is_weighted>>
+{
+protected:
+    using State = AggregateFunctionGroupSortedArrayData<T, is_weighted>;
+    using Base = IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>;
+
+    UInt64 threshold;
+    DataTypePtr & input_data_type;
+    mutable std::mutex mutex;
+
+    static void deserializeAndInsert(StringRef str, IColumn & data_to);
+
+public:
+    AggregateFunctionGroupSortedArray(UInt64 threshold_, const DataTypes & argument_types_, const Array & params)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>(
+            argument_types_, params)
+        , threshold(threshold_)
+        , input_data_type(this->argument_types[0])
+    {
+    }
+
+    void create(AggregateDataPtr place) const override
+    {
+        Base::create(place);
+        this->data(place).threshold = threshold;
+    }
+
+    String getName() const override { return is_weighted ? "groupSortedArrayWeighted" : "groupSortedArray"; }
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeArray>(input_data_type); }
+
+    bool allocatesMemoryInArena() const override
+    {
+        if constexpr (std::is_same_v<T, StringRef>)
+            return true;
+        else
+            return false;
+    }
+
+    void add(AggregateDataPtr __restrict place, const IColumn **columns, size_t row_num, Arena *arena) const override { 
+        State & data = this->data(place);
+        if constexpr (is_weighted)
+        {
+            data.add(readItem<T, is_plain_column>(columns[0], arena, row_num),
+                     readItem<UInt64, false>(columns[1], arena, row_num));
+        }
+        else
+        {
+             data.add(readItem<T, is_plain_column>(columns[0], arena, row_num));
+        }
+    }
+
+    void addBatchSinglePlace(
+        size_t batch_size, AggregateDataPtr place, const IColumn ** columns, Arena * arena, ssize_t if_argument_pos) const override
+    {
+        State & data = this->data(place);
+        if constexpr (is_weighted)
+        {
+            StringRef ref = columns[1]->getRawData();
+            UInt64 values[batch_size];
+            memcpy(values, ref.data, batch_size * sizeof(UInt64));
+            UInt64 * bestRows = new UInt64[this->threshold];
```

Comment:
Please, no new and no delete[]. No-no-no-no.no-no :D Either use a VLA like you did with values or put it inside a smart pointer if you want it to go to the heap.